aten/src/ATen/native/special_functions/detail/promote.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+namespace at {
+namespace native {
+namespace special_functions {
+namespace detail {
+template<typename T1, bool = std::is_integral<T1>::value>
+struct promotion {
+  using type = double;
+};
+
+template<typename T1>
+struct promotion<T1, false> {};
+
+template<>
+struct promotion<float> {
+  using type = float;
+};
+
+template<>
+struct promotion<double> {
+  using type = double;
+};
+
+template<>
+struct promotion<long double> {
+  using type = long double;
+};
+
+template<typename... T>
+using promotion_t = typename promotion<T...>::type;
+
+template<typename T1>
+struct promotion<c10::complex<T1>, false> {
+ private:
+  using value_type = typename c10::complex<T1>::value_type;
+ public:
+  using type = decltype(c10::complex<promotion_t<value_type>>{});
+};
+
+template<typename T1, typename... T>
```

Comment:
currently all uses of promote don't have varargs, do we need it here? (and if they did, type promotion rules in pytorch are more complex and depend on dimensionality of tensors)

Reply:
Good point! I don’t think think is needed.